R/helpers.R

```diff
@@ -186,37 +186,43 @@ extract_num_iterations <- function(x) {
 # the model was trained with the `valids` parameter set such that error metrics
 # are saved for each tree on the model$record_evals attribute. The output
 # weights are useful for computing comps using leaf node assignments
-extract_tree_weights <- function(model, init_score, metric = "rmse") {
-  # Index into the errors list, and un-list so it is a flat/1dim list
-  record_evals <- model$record_evals
-  errors <- unlist(record_evals$tree_errors[[metric]]$eval)
-  errors <- c(init_score, errors)
-  diff_in_errors <- diff(errors, 1, 1)
+extract_tree_weights <- function(model,
+                                 init_score,
+                                 training_data,
+                                 outcome,
+                                 num_iterations) {
+  tree_predictions <- matrix(
+    nrow = nrow(training_data),
+    ncol = num_iterations + 1
+  )
+  tree_predictions[, 1] <- init_score
+  for (num_iteration in seq_len(num_iterations)) {
+    if (num_iteration %% 100 == 0) {
+      message(
+        glue::glue("Extracting weights for tree {num_iteration}")
+      )
+    }
+    tree_predictions[, num_iteration + 1] <- predict(
+      object = model,
+      newdata = as.matrix(training_data),
+      num_iteration = num_iteration
+    )
+  }
+  tree_errors <- abs(outcome - tree_predictions)
+  # Get a lagged diff of the error matrix by row. Since diff() operates on
+  # matrix columns, we need to transpose the input
+  diff_in_errors <- tree_errors %>%
+    t() %>%
+    diff(1, 1) %>%
+    t() * -1
```

Comment:
@jeancochrane The tree weights calc here is the correct version. We're starting with the mean sale price from the training data as the initial score, then calculating the diff for each subsequent iteration for each observation.